Match ergonomics 2024: `mut` doesn't reset binding mode

Jules-Bertholet opened the pull request:
r? @Nadrieril

cc https://github.com/rust-lang/rust/issues/123076

@rustbot label A-edition-2024 A-patterns

Jules-Bertholet committed:
Match ergonomics 2024: `mut` doesn't reset binding mode

Jules-Bertholet committed:
Migration lint

Rustfix remains TODO

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8578433287/job/23512269999) [(plain)](https://github.com/rust-lang/rust/commit/aa4ffd80f8492ad4d6f1031a29c77d278a76d420/checks/23512269999/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 29.0s done
#16 DONE 40.1s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
+    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+    |
+    = note: the `dereferencing_mut_binding` lint is unstable
+    = note: see issue #123076 <https://github.com/rust-lang/rust/issues/123076> for more information
+    = help: add `#![feature(mut_dont_reset_binding_mode_2024)]` to the crate attributes to enable
+    = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
+ 
+ 
1 error: call to unsafe function `unsf` is unsafe and requires unsafe block (error E0133)
3    |

18    |         ^^^^^^^^^^^^^^^^^^^^^^^
19    = note: `#[deny(unsafe_op_in_unsafe_fn)]` implied by `#[deny(rust_2024_compatibility)]`
---
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: the `dereferencing_mut_binding` lint is unstable
   = note: see issue #123076 <https://github.com/rust-lang/rust/issues/123076> for more information
   = help: add `#![feature(mut_dont_reset_binding_mode_2024)]` to the crate attributes to enable
   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date


error: call to unsafe function `unsf` is unsafe and requires unsafe block (error E0133)
   |
LL |     unsf();
   |     ^^^^^^ call to unsafe function
   |
```

</details>

Jules-Bertholet committed:
Temporarily remove future compatibility label from migration lint

The lint is unstable, and the lint group `rust_2024_compatibility` must keep working on stable

Nadrieril commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -629,12 +629,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                BindingAnnotation(ByRef::No, Mutability::Mut)
+            }
+            BindingAnnotation(ByRef::No, mutbl) => BindingAnnotation(def_br, mutbl),
+            BindingAnnotation(ByRef::Yes(_), _) => ba,
```
This doesn't preserve old behavior in the case where `ba` is `BindingAnnotation(ByRef::No, Mutability::Not)` and `def_bm.1` is `Mutability::Mut`. I guess that can't happen, right? Could we store just a `ByRef` in `PatInfo`?

Jules-Bertholet replied:
> Could we store just a `ByRef` in `PatInfo`?

I think we can, but it would require changes to `calc_default_binding_mode`, so I'll do it in a later PR so as not to conflict with #123512

Nadrieril reviewed (approved):
LGTM modulo one detail and one question

EDIT: oops I was only looking at the first commit

Nadrieril commented on compiler/rustc_lint_defs/src/builtin.rs:
```diff
@@ -1626,6 +1627,41 @@ declare_lint! {
+    /// ### Explanation
+    ///
+    /// Without the `mut`, `x` would have type `&u32`. Pre-2024, adding `mut` makes `x` have type
+    /// `u32`, which was deeped surprising. After edition 2024, adding `mut` will not change the
```
```suggestion
    /// `u32`, which was deemed surprising. After edition 2024, adding `mut` will not change the
```

Nadrieril reviewed (approved):
Now I've reviewed all commits :+1:

Jules-Bertholet committed:
Fix typo

Co-authored-by: Guillaume Boisseau <Nadrieril@users.noreply.github.com>

Jules-Bertholet committed:
Rename feature gate

Jules-Bertholet committed:
Mark gate as incomplete

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8680348466/job/23800702727) [(plain)](https://github.com/rust-lang/rust/commit/568d82a0b74379a4a14594e8118693e1876f2273/checks/23800702727/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 29.5s done
#16 DONE 32.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---

---- [ui] tests/ui/pattern/mut_preserve_binding_mode_2021.rs stdout ----
diff of stderr:

+ warning: the feature `mut_preserve_binding_mode_2024` is incomplete and may not be safe to use and/or cause compiler crashes
+    |
+    |
+ LL | #![feature(mut_preserve_binding_mode_2024)]
+    |
+    = note: see issue #123076 <https://github.com/rust-lang/rust/issues/123076> for more information
+    = note: `#[warn(incomplete_features)]` on by default
+ 
---
32 


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/pattern/mut_preserve_binding_mode_2021/mut_preserve_binding_mode_2021.stderr
To only update this specific test, also pass `--test-args pattern/mut_preserve_binding_mode_2021.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/pattern/mut_preserve_binding_mode_2021.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/pattern/mut_preserve_binding_mode_2021" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/pattern/mut_preserve_binding_mode_2021/auxiliary" "--edition=2021" "-Zunstable-options"
--- stderr -------------------------------
--- stderr -------------------------------
warning: the feature `mut_preserve_binding_mode_2024` is incomplete and may not be safe to use and/or cause compiler crashes
   |
LL | #![feature(mut_preserve_binding_mode_2024)]
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   |
   = note: see issue #123076 <https://github.com/rust-lang/rust/issues/123076> for more information
   = note: `#[warn(incomplete_features)]` on by default

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/pattern/mut_preserve_binding_mode_2021.rs:9:9
   |
LL |     let Foo(mut a) = &Foo(0);
LL |     a = &42;
LL |     a = &42;
   |         ^^^ expected `u8`, found `&{integer}`
help: consider removing the borrow
   |
LL -     a = &42;
LL +     a = 42;
LL +     a = 42;
   |

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/pattern/mut_preserve_binding_mode_2021.rs:13:9
   |
LL |     let Foo(mut a) = &mut Foo(0);
LL |     a = &mut 42;
LL |     a = &mut 42;
   |         ^^^^^^^ expected `u8`, found `&mut {integer}`
help: consider removing the borrow
   |
LL -     a = &mut 42;
LL +     a = 42;
---

---- [ui] tests/ui/pattern/mut_preserve_binding_mode_2024_lint.rs stdout ----
diff of stderr:

+ warning: the feature `mut_preserve_binding_mode_2024` is incomplete and may not be safe to use and/or cause compiler crashes
+   --> $DIR/mut_preserve_binding_mode_2024_lint.rs:2:12
+    |
+ LL | #![feature(mut_preserve_binding_mode_2024)]
+    |
+    = note: see issue #123076 <https://github.com/rust-lang/rust/issues/123076> for more information
+    = note: `#[warn(incomplete_features)]` on by default
+ 
+ 
1 error: dereferencing `mut` binding
2   --> $DIR/mut_preserve_binding_mode_2024_lint.rs:9:13
3    |

27 LL |     let Foo(mut a) = &mut Foo(0);
29 
- error: aborting due to 2 previous errors
+ error: aborting due to 2 previous errors; 1 warning emitted
31 
---
To only update this specific test, also pass `--test-args pattern/mut_preserve_binding_mode_2024_lint.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/pattern/mut_preserve_binding_mode_2024_lint.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/pattern/mut_preserve_binding_mode_2024_lint" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/pattern/mut_preserve_binding_mode_2024_lint/auxiliary" "--edition=2021"
--- stderr -------------------------------
--- stderr -------------------------------
warning: the feature `mut_preserve_binding_mode_2024` is incomplete and may not be safe to use and/or cause compiler crashes
##[warning]  --> /checkout/tests/ui/pattern/mut_preserve_binding_mode_2024_lint.rs:2:12
LL | #![feature(mut_preserve_binding_mode_2024)]
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #123076 <https://github.com/rust-lang/rust/issues/123076> for more information
   = note: see issue #123076 <https://github.com/rust-lang/rust/issues/123076> for more information
   = note: `#[warn(incomplete_features)]` on by default

error: dereferencing `mut` binding
##[error]  --> /checkout/tests/ui/pattern/mut_preserve_binding_mode_2024_lint.rs:9:13
   |
LL |     let Foo(mut a) = &Foo(0);
   |             ^^^^^ `mut` dereferences the type of this binding
help: this will change in edition 2024
  --> /checkout/tests/ui/pattern/mut_preserve_binding_mode_2024_lint.rs:9:13
   |
   |
LL |     let Foo(mut a) = &Foo(0);
note: the lint level is defined here
  --> /checkout/tests/ui/pattern/mut_preserve_binding_mode_2024_lint.rs:4:11
   |
   |
LL | #![forbid(dereferencing_mut_binding)]

error: dereferencing `mut` binding
##[error]  --> /checkout/tests/ui/pattern/mut_preserve_binding_mode_2024_lint.rs:13:13
   |
   |
LL |     let Foo(mut a) = &mut Foo(0);
   |             ^^^^^ `mut` dereferences the type of this binding
help: this will change in edition 2024
  --> /checkout/tests/ui/pattern/mut_preserve_binding_mode_2024_lint.rs:13:13
   |
   |
LL |     let Foo(mut a) = &mut Foo(0);

error: aborting due to 2 previous errors; 1 warning emitted
------------------------------------------



---- [ui] tests/ui/pattern/mut_preserve_binding_mode_2024.rs stdout ----
normalized stderr:
warning: the feature `mut_preserve_binding_mode_2024` is incomplete and may not be safe to use and/or cause compiler crashes
   |
LL | #![feature(mut_preserve_binding_mode_2024)]
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
---



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/pattern/mut_preserve_binding_mode_2024/mut_preserve_binding_mode_2024.stderr
To only update this specific test, also pass `--test-args pattern/mut_preserve_binding_mode_2024.rs`

error: 1 errors occurred comparing output.
status: exit status: 0
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/pattern/mut_preserve_binding_mode_2024.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-O" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/pattern/mut_preserve_binding_mode_2024/a" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/pattern/mut_preserve_binding_mode_2024/auxiliary" "--edition=2024" "-Zunstable-options"
--- stderr -------------------------------
--- stderr -------------------------------
warning: the feature `mut_preserve_binding_mode_2024` is incomplete and may not be safe to use and/or cause compiler crashes
   |
LL | #![feature(mut_preserve_binding_mode_2024)]
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
```

</details>

Jules-Bertholet committed:
Fix tests

bors commented:
:umbrella: The latest upstream changes (presumably #123991) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Nadrieril commented:
@bors r+

bors commented:
:pushpin: Commit 7a3211726bb101ab0c0138b2dcff44a175ff74c4 has been approved by `Nadrieril`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nadrieril 7a3211726bb101ab0c0138b2dcff44a175ff74c4 -->
<!-- homu: {"type":"Approved","sha":"7a3211726bb101ab0c0138b2dcff44a175ff74c4","approver":"Nadrieril","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
